src/Npgsql/NpgsqlConnector.cs

```diff
@@ -936,11 +916,48 @@ await socket.ConnectAsync(endpoint)
                     if (i == endpoints.Length - 1)
                         throw new NpgsqlException($"Failed to connect to {endpoint}", e);
                 }
-                finally
+            }
+
+            Task<IPAddress[]> GetHostAddressesAsync(NpgsqlTimeout timeout, CancellationToken cancellationToken)
+            {
+                // .NET 6.0 added cancellation support to GetHostAddressesAsync on Windows, which allows us to implement real
+                // cancellation and timeout. On older TFMs, we fake-cancel the operation, i.e. stop waiting
+                // and raise the exception, but the actual connection task is left running.
+#if NET6_0_OR_GREATER
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```

Comment:
Do we really want to encode assumptions in our source as to which platforms support DNS cancellation and which don't? Who knows what will happen in the future... We could just always have the WithCancellationAndTimeout mechanism wrapping the call, and on Windows the two mechanisms would be active.

If we *do* want to differentiate Windows vs. non-Windows (again I don't think we should), then the two invocations below are basically duplicates and can be de-duplicated.

Reply:
Agreed. I hope we have finally found the good combination here.
Comments are welcome indeed :-)